waffle_hub/hub/hub.py

```diff
@@ -1287,29 +1289,43 @@ def inferencing(self, cfg: InferenceConfig, callback: InferenceCallback) -> str:
                 results.append({str(image_info.image_rel_path): [res.to_dict() for res in result]})
 
                 if cfg.draw:
+                    io.make_directory(self.draw_dir)
                     draw = draw_results(
                         image_info.ori_image,
                         result,
                         names=[x["name"] for x in self.categories],
                     )
-                    draw_path = self.draw_dir / Path(image_info.image_rel_path).with_suffix(".png")
-                    io.make_directory(draw_path.parent)
-                    cv2.imwrite(str(draw_path), draw)
-
-                if cfg.draw and cfg.source_type == "video":
-                    if writer is None:
-                        h, w = draw.shape[:2]
-                        writer = cv2.VideoWriter(
-                            str(self.inference_dir / Path(cfg.source).with_suffix(".mp4").name),
-                            cv2.VideoWriter_fourcc(*"mp4v"),
-                            dataset.fps,
-                            (w, h),
+
+                    if cfg.source_type == "video":
+                        if writer is None:
+                            h, w = draw.shape[:2]
+                            writer = create_video_writer(
+                                str(self.inference_dir / Path(cfg.source).with_suffix(".mp4").name),
+                                dataset.fps,
+                                (w, h),
+                            )
+                        writer.write(draw)
+
+                        draw_path = (
+                            self.draw_dir
+                            / Path(cfg.source).stem
+                            / Path(image_info.image_rel_path).with_suffix(".png")
+                        )
+                    else:
+                        draw_path = self.draw_dir / Path(image_info.image_rel_path).with_suffix(
+                            ".png"
                         )
-                    writer.write(draw)
+                    save_image(draw_path, draw, create_directory=True)
 
                 if cfg.show:
+                    if not cfg.draw:
+                        draw = draw_results(
+                            image_info.ori_image,
+                            result,
+                            names=[x["name"] for x in self.categories],
+                        )
                     cv2.imshow("result", draw)
-                    cv2.waitKey(0)
+                    cv2.waitKey(30)
```

Comment:
It will slow down the inference.
30ms will emit 33fps, but in inference we need to consider the inference time.
So, we usually set sleep time to 1ms. 0ms was my mistake 😂

Reply:
i'll fix it and merge.